From dc4c470f94712dacd9dbd7562798e642f3003eb7 Mon Sep 17 00:00:00 2001
From: Vitaly Dyachkov <vitaly@igalia.com>
Date: Fri, 21 Apr 2023 12:43:32 -0700
Subject: [PATCH] [WPE][GTK] Do not deref CreationData's buffer when Cairo font
 is destroyed https://bugs.webkit.org/show_bug.cgi?id=255785

Reviewed by Carlos Alberto Lopez Perez.

Since 263084@main `FontCustomPlatformData` takes
`FontPlatformData::CreationData` instead of `FragmentedSharedBuffer`.
The buffer is now a reference counted field of
`FontPlatformData::CreationData` struct and will be automatically
dereferenced when `FontCustomPlatformData` is destroyed.
It is an error to manually dereference it using Cairo font destroy
function.

* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::FontCustomPlatformData):
(WebCore::releaseCustomFontData): Deleted.

Canonical link: https://commits.webkit.org/263258@main
---
 Source/WebCore/platform/graphics/Font.h                |  1 -
 .../WebCore/platform/graphics/FontCustomPlatformData.h |  1 -
 .../freetype/FontCustomPlatformDataFreeType.cpp        | 10 ----------
 3 files changed, 12 deletions(-)

diff --git a/Source/WebCore/platform/graphics/Font.h b/Source/WebCore/platform/graphics/Font.h
index b69f433adbd6..39508e4252db 100644
--- a/Source/WebCore/platform/graphics/Font.h
+++ b/Source/WebCore/platform/graphics/Font.h
@@ -61,7 +61,6 @@ namespace WebCore {
 class FontCache;
 class FontDescription;
 class GlyphPage;
-class FragmentedSharedBuffer;
 
 struct GlyphData;
 
diff --git a/Source/WebCore/platform/graphics/FontCustomPlatformData.h b/Source/WebCore/platform/graphics/FontCustomPlatformData.h
index aa12d82532d7..c551a9e93ae7 100644
--- a/Source/WebCore/platform/graphics/FontCustomPlatformData.h
+++ b/Source/WebCore/platform/graphics/FontCustomPlatformData.h
@@ -49,7 +49,6 @@ namespace WebCore {
 class SharedBuffer;
 class FontDescription;
 class FontCreationContext;
-class FragmentedSharedBuffer;
 
 template <typename T> class FontTaggedSettings;
 typedef FontTaggedSettings<int> FontFeatureSettings;
diff --git a/Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp b/Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp
index e7857fe1818f..38daa2b7ff24 100644
--- a/Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp
+++ b/Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp
@@ -38,11 +38,6 @@
 
 namespace WebCore {
 
-static void releaseCustomFontData(void* data)
-{
-    static_cast<FragmentedSharedBuffer*>(data)->deref();
-}
-
 static cairo_user_data_key_t freeTypeFaceKey;
 
 FontCustomPlatformData::FontCustomPlatformData(FT_Face freeTypeFace, FontPlatformData::CreationData&& data)
@@ -50,11 +45,6 @@ FontCustomPlatformData::FontCustomPlatformData(FT_Face freeTypeFace, FontPlatfor
     , creationData(WTFMove(data))
     , m_renderingResourceIdentifier(RenderingResourceIdentifier::generate())
 {
-    static cairo_user_data_key_t bufferKey;
-    auto buffer = creationData.fontFaceData;
-    cairo_font_face_set_user_data(m_fontFace.get(), &bufferKey, &buffer,
-         static_cast<cairo_destroy_func_t>(releaseCustomFontData));
-
     // Cairo doesn't do FreeType reference counting, so we need to ensure that when
     // this cairo_font_face_t is destroyed, it cleans up the FreeType face as well.
     cairo_font_face_set_user_data(m_fontFace.get(), &freeTypeFaceKey, freeTypeFace,
